boden/grund.cc

```diff
@@ -1883,94 +1883,88 @@ bool grund_t::weg_erweitern(waytype_t wegtyp, ribi_t::ribi ribi)
 	}
 	return false;
 }
+struct {
+	static void visit(baum_t* tree){delete tree;}
+	static void visit(groundobj_t* groundobj){delete groundobj;}
+} tree_groundobj_remove_visitor;
+struct {
+	static void visit(const baum_t*){}
+	static void visit(const groundobj_t*){}
+} tree_groundobj_noop_visitor;
+
+template<typename this_T, typename visitor_T>
+sint64 grund_t::visit_trees_and_groundobjs(this_T& self, visitor_T& visitor) {
+	sint64 costs=0;
+	for(  uint8 i=0;  i<self.get_top();  i++  ) {
+		obj_t *obj = self.obj_bei(i);
+		switch(obj->get_typ()) {
+			case obj_t::baum: {
+				baum_t* const tree = (baum_t *)obj;
+				visitor.visit(tree);
+				costs -= welt->get_settings().cst_remove_tree;
+				break;
+			}
+			case obj_t::groundobj: {
+				groundobj_t* gr_obj = (groundobj_t *) obj;
+				costs += gr_obj->get_desc()->get_value();
+				visitor.visit(gr_obj);
+				break;
+			}
+			default: break;
+		}
+	}
+	return costs;
+}
 
-/**
- * remove trees and groundobjs on this tile
- * called before building way or powerline
- * @return costs
- */
-sint64 grund_t::remove_trees()
-{
-	sint64 cost=0;
-	// remove all trees ...
-	while (baum_t* const d = find<baum_t>(0)) {
-		// we must mark it by hand, since we want to join costs
-		d->mark_image_dirty( get_image(), 0 );
-		delete d;
-		cost -= welt->get_settings().cst_remove_tree;
-	}
-	// remove all groundobjs ...
-	while (groundobj_t* const d = find<groundobj_t>(0)) {
-		cost += d->get_desc()->get_value();
-		delete d;
-	}
-	return cost;
+sint64 grund_t::get_tree_remove_costs() const {
+	return visit_trees_and_groundobjs(*this, tree_groundobj_noop_visitor);
 }
 
+sint64 grund_t::remove_trees() {
+	return visit_trees_and_groundobjs(*this, tree_groundobj_remove_visitor);
+}
```

Comment:
Share code to calculate removal costs of tree and groundobjs with the actual remove function.
The only difference is constnes and wether the delete is called or not.